README.md
# lightstep-tracer

> ❗ **This instrumentation is no longer recommended**. Please review [documentation on setting up and configuring the OpenTelemetry Node.js Launcher](https://github.com/lightstep/otel-launcher-node) or [OpenTelemetry JS (Browser)](https://github.com/open-telemetry/opentelemetry-js) for more information on migrating to OpenTelemetry.

[![npm version](https://badge.fury.io/js/lightstep-tracer.svg)](https://badge.fury.io/js/lightstep-tracer)
[![Circle CI](https://circleci.com/gh/lightstep/lightstep-tracer-javascript.svg?style=shield)](https://circleci.com/gh/lightstep/lightstep-tracer-javascript)
[![MIT license](http://img.shields.io/badge/license-MIT-blue.svg)](http://opensource.org/licenses/MIT)

LightStep distributed tracing library for Node.js and the browser.

## Installation

```
npm install --save lightstep-tracer opentracing
```

All modern browsers and Node versions >= 12 are supported. Users who need support for Node 8 and 10 can use v0.32.0 and prior.

## Getting started

To use LightStep as the OpenTracing binding, initialize the global `Tracer` with the LightStep implementation:

```javascript
var opentracing = require('opentracing');
var lightstep   = require('lightstep-tracer');

opentracing.initGlobalTracer(new lightstep.Tracer({
    access_token   : '{your_access_token}',
    component_name : '{your_service_or_app_name}',
}));
```

The **[LightStep JavaScript Tracing Cookbook](doc/cookbook.md)** is a good next stop for information on how to quickly instrument your system.  If you want to try something out quickly in your browser code, see the **[browser quick start example](doc/cookbook.md#browser-quick-start)**.

* For more information about using the OpenTracing API, see http://opentracing.io/
* See [examples/browser](https://github.com/lightstep/lightstep-tracer-javascript/tree/master/examples/browser) for a complete JavaScript browser example
* See [examples/node](https://github.com/lightstep/lightstep-tracer-javascript/tree/master/examples/node) for a complete Node.js server-side example

The browser version of the code can be explicitly included using the following, which can be helpful in some [`browserify`](https://github.com/substack/node-browserify) (or similar) setups:

```javascript
var lightstep = require('lightstep-tracer/browser');
```


## LightStep-specific API

The OpenTracing standard JavaScript API is [documented here](https://doc.esdoc.org/github.com/opentracing/opentracing-javascript/). The following describes LightStep-specific options and methods.

### LightStep

---

#### new Tracer(options)

**Required options**

* `access_token` `string` *required* - the project access token
* `component_name` `string` *required* - the string identifier for the application, service, or process

**Standard options**

* `verbosity` `number` *optional, default=1* - controls the level of logging to the console
    - `0` - the client library will *never* log to the console
    - `1` - error reporting will be throttled to the first error per minute
    - `2` - all errors are logged to the console
    - `3` - all errors, warnings, and info statements are logged to the console
    - `4` - all log statements, including debugging details
* `collector_host` `string` *optional* - custom collector hostname
* `collector_port` `number` *optional* - custom collector port
* `collector_path` `string` *optional* - custom collector base path (if served behind a reverse proxy)
* `collector_encryption` `string` *optional, default='tls'*
    - `tls` - use HTTPS encrypted connections
    - `none` - use HTTP plain-text connections

**Browser-specific initialization options**

* `instrument_page_load` `bool` - creates a long-lived single span for the entire page view and is recommended for short-page visits in a multi-page website. For a single-page web app, this behavior may be undesirable. Defaults to false. This must be set at initialization, changes after initialization will have no effect.

* `xhr_instrumentation` `bool` - if false, disables automatic instrumentation of XMLHttpRequests (XHRs). This must be set at initialization; changes after initialization will have no effect. Defaults to false.

* `xhr_url_inclusion_patterns` `RegExp[]` - an array of regular expressions used to whitelist URLs for `XMLHttpRequest` auto-instrumentation. The default value is wildcard matching all strings. For a given URL to be instrumented, it must match at least one regular expression in `xhr_url_inclusion_patterns` and not match any regular expressions in `xhr_url_exclusion_patterns`.

* `xhr_url_exclusion_patterns` `RegExp[]` - an array of regular expressions used to exclude URLs from `XMLHttpRequest` auto-instrumentation. The default value is an empty array. For a given URL to be instrumented, it must match at least one regular expression in `xhr_url_inclusion_patterns` and not match any regular expressions in `xhr_url_exclusion_patterns`.

* `xhr_url_header_inclusion_patterns` `RegExp[]` - an array of regular expressions used to include URLs which auto-instrumented `XMLHttpRequest`s add tracing headers to. The default value is wildcard matching all strings. For a given URL to have tracing headers added, it must match at least one regular expression in `xhr_url_header_inclusion_patterns` and not match any regular expressions in `xhr_url_header_exclusion_patterns`.

* `xhr_url_header_exclusion_patterns` `RegExp[]` - an array of regular expressions used to exclude URLs that auto-instrumented `XMLHttpRequest`s add tracing headers to. The default value is an empty array. For a given URL to have tracing headers added, it must match at least one regular expression in `xhr_url_header_inclusion_patterns` and not match any regular expressions in `xhr_url_header_exclusion_patterns`.

* `fetch_instrumentation` `bool` - if false, disables automatic instrumentation of `window.fetch`. This must be set at initialization; changes after initialization will have no effect. Defaults to false.

* `fetch_url_inclusion_patterns` `RegExp[]` - an array of regular expressions used to whitelist URLs for `window.fetch` auto-instrumentation. The default value is wildcard matching all strings. For a given URL to be instrumented, it must match at least one regular expression in `fetch_url_inclusion_patterns` and not match any regular expressions in `fetch_url_exclusion_patterns`.

* `fetch_url_exclusion_patterns` `RegExp[]` - an array of regular expressions used to exclude URLs from `window.fetch` auto-instrumentation. The default value is an empty array. For a given URL to be instrumented, it must match at least one regular expression in `fetch_url_inclusion_patterns` and not match any regular expressions in `fetch_url_exclusion_patterns`.

* `fetch_url_header_inclusion_patterns` `RegExp[]` - an array of regular expressions used to include URLs for `window.fetch` tracer header inclusion. The default value is wildcard matching all strings. For a given URL to be have tracing headers added, it must match at least one regular expression in `fetch_url_header_inclusion_patterns` and not match any regular expressions in `fetch_url_header_exclusion_patterns`.

* `fetch_url_header_exclusion_patterns` `RegExp[]` - an array of regular expressions used to exclude header insertion onto URLs from `window.fetch` auto-instrumentation. The default value is an empty array. For a given URL to have tracing headers added, it must match at least one regular expression in `fetch_url_header_inclusion_patterns` and not match any regular expressions in `fetch_url_header_exclusion_patterns`.

* `include_cookies` `bool` - if true, includes cookies in the span logs for both `window.fetch` and `XMLHttpRequest`. Defaults to true.

**node-specific initialization options**

* `nodejs_instrumentation` `bool` - if false, disables automatic instrumentation of node requests.
This must be set at initialization; changes after initialization will have no effect. Defaults to false.

* `nodejs_url_inclusion_patterns` `RegExp[]` - an array of regular expressions used to whitelist
URLs
for the `http` and `https` modules auto-instrumentation. The default value is wildcard matching all strings.
For a given URL to be instrumented, it must match at least one regular expression in
`nodejs_url_inclusion_patterns` and not match any regular expressions in
`nodejs_url_exclusion_patterns`.

* `nodejs_url_exclusion_patterns` `RegExp[]` - an array of regular expressions used to exclude URLs
from the `http` and `https` modules auto-instrumentation. The default value is an empty array. For a given URL
to be instrumented, it must match at least one regular expression in
`nodejs_url_inclusion_patterns`
and not match any regular expressions in `nodejs_url_exclusion_patterns`.

**Non-standard options**

*NOTE: Future API compatibility on non-standard options is not guaranteed.*

* `disable_reporting_loop` `bool` *optional*, *default=false* - if true, the timer that automatically sends reports to the collector will be disabled. This option is independent of `disable_report_on_exit`.
* `disable_report_on_exit` `bool` *optional*, *default=false* - if true, the final report that is automatically sent at process exit in Node or page unload in the browser will not be sent.
* `report_timeout_millis` `number` *optional*, *default=30000* - the default timeout value, in milliseconds, for reports to the LightStep collector
* `gzip_json_requests` `bool` *optional*, *default=true* - if true, the reports will be gzipped before sent to the collector.
* `default_span_tags` `string` *optional* - an associative array of tags to add to every span started by the tracer (e.g., the active user id in a browser client)
* `delay_initial_report_millis` `int` *optional*, *default=1000* - maximum additional delay of the initial report in addition to the normal reporting interval. A value between zero and this maximum will be selected as the actual delay. This can be useful when concurrently launching a large number of new processes and there is a desire to distribute the initial reports over a window of time.
* `error_throttle_millis` `int` *optional*, *default=60000* - when `verbosity` is set to `1`, this the minimum time between logged errors.
* `transport` `string` *optional*, *default=proto* - when `transport` is set to `thrift`, the Tracer will use Thrift as its transport instead of Proto over HTTP. (Not supported in React-Native)
* `logger` `function(level: string, message: string, payload: any): void` *optional* - specify a custom logger function. Possible `level` values are `debug`, `info`, `warn` and `error`. By default messages will be logged to the console.
* `disable_meta_event_reporting` `bool` *optional*, *default=false* - when `disable_meta_event_reporting` is set to `true`, the tracer will disable meta event reporting even if requested by the Satellite.
* `propagators` `dictionary` *optional*, *defaults=*`{ [opentracing.FORMAT_HTTP_HEADERS]: new lightstep.LightStepPropagator(), [opentracing.FORMAT_TEXT_MAP]: new lightstep.LightStepPropagator(), [opentracing.FORMAT_BINARY]: new lightstep.UnsupportedPropagator() }`: Allows inject/extract to use custom propagators for different formats. This package includes `B3Propagator` that supports B3 headers on text maps and http headers. `DDPropagator` supports DataDog trace headers.
* `clear_span_buffer_consecutive_errors` `number` *optional*, *default=null* - each consecutive buffer flush error will check to see if `clear_span_buffer_consecutive_errors` has been reached.  If reached, the span buffer will be emptied.  This is useful for auto-recovering from errors based on request size constraints, particularly max payload size on intermediate load balancers. 

An example configuration using custom propagators might look like:
```js
const tracer = new lightstep.Tracer({
  propagators: {
    [opentracing.FORMAT_HTTP_HEADERS]: new lightstep.DDPropagator(),
    [opentracing.FORMAT_TEXT_MAP]: new lightstep.B3Propagator()
  }
})
```

### SpanImp

---

#### generateTraceURL()

Returns an absolute URL to the LightStep application for the trace containing this span. It is safe to call this method after `finish()`.

```js
...
span.finish();

var url = span.generateTraceURL())
console.log('View the trace for this span at:', url);
```

### Release process using circleCI 
1. Run command (patch, minor, major)
```commandline
make release RELEASE_TYPE=patch
```

### Release process using circleCI with custom version 
1. Run command 
```commandline
make release RELEASE_TYPE=v3.2.1-no-protobuf
```
2. It will create a tag "v3.2.1-no-protobuf" for version "3.2.1"


## License

[The MIT License](LICENSE).

Copyright (c) 2016, LightStep


CHANGELOG.md
# CHANGELOG

*Log of significant changes, especially those affecting the supported API.*

## vNext

## 0.35.0
* Upgrade webpack (#304)
* Upgrade cacheable-request, package-json (#303)
* Upgrade http-cache-semantics (#302)
* Upgrade json5 (#301)
* Upgrade loader-utils (#299)
* Upgrade opentracing (#298)
* Upgrade thrift (#297)
* Upgrade async (#296)
* Upgrade google-protobuf (#295)
* Upgrade terser (#293)

## 0.34.0
* Omit non-sampled spans from report (#290)

## 0.33.0
* Upgrade async (#288)
* Upgrade to Babel 7 (#288)
* Upgrade various dependecies (#288)
* Substitute deprecated packages (#288)
* add Node 14 and Node 16 jobs on CI (#288)
* Drop support for Node 8 and 10 (#288)

## 0.32.0
* Add missing OpenTracing spancontext methods (#279)
* Update ws and normalize-url dependencies.

## 0.31.2
* Update to thrift v0.14.1

## 0.31.1
* Update versions of underscore, lodash, handlebars, and y18n.

## 0.31.0
* Improve performance of hexToDec by using native BigInt when available.

## 0.30.2
* Bugfix for nodejs: remove throw in fetch onError handler (#246)

## 0.30.1
* Fix misnamed configuration options. `nodejs_inclusion_patterns` was renamed to `nodejs_url_inclusion_patterns` and `nodejs_exclusion_patterns` was renamed to `nodejs_url_exclusion_patterns`. Because of this bug, the inclusion and exclusion patterns were not being recognized by instrumentation.

## 0.30.0

* Bug fix for eslint for Microsoft Edge (#241)
* Add options for auto-instrumented urls and tracing headers (#237)

## 0.29.0

* Truncate most significant 64 bits when all 0's in B3 propagator.Inject (#238)
* Support React Native with the browser tracer (#235)

## 0.28.0

* Bug fix for propagating baggage from parent span to child span
* SpanContext is now exported to facilitate creating custom propagators

## 0.27.0

* No changes

## 0.26.0

* Require node >= v8
* New option to empty span buffer on max error streak `clear_span_buffer_consecutive_errors`
* Upgrade eslint

## 0.25.2

* fix for fetch plugin: 'method' was always GET
* Fix missing clock offsets over proto transport
* Fix timestamp conversion for protobuf transport

## 0.25.1
* No changes (minor bump to align with 0.25.1-no-protobuf release)

## 0.25.0
* Bug fix on `fetch` instrumentation options headers being overwritten in some cases.
* Re-build dist to fix failed page load on IE11 (#207)

## 0.24.3
* Fix how the `fetch` shim so that it does not overwrite user-provided `Response.headers`
* Add clock correction and error handling for the protobuf transport's responses
* Specify correct modules in node instrumentation docs
* Add a CONTRIBUTING.md guide

## 0.24.2
* Upgrade thrift to 0.13.0 (#198)

## 0.24.1
* Guard session storage access
* update type of collector_encryption
* Fix propagator typo
* Update typing on inclusion/exclusion patterns to match docs

## 0.24.0
* Adds DataDog header support (#187)
* Removes for..of loop to support IE11 (#189)

## 0.23.0

* Add NodeJS instrumentation and shims (#184)
* Adds B3 propagator (#177)

## 0.22.3
* Fixes an issue where reporter cookies would be set with invalid key strings.

## 0.22.2
* Completely defines the `TracerOptions` interface.
* Updates `thrift` package to address a security vulnerability.

## 0.22.1
* Fixes a bug where certain tracer tags were not set when using `proto` transport.
* Adds TypeScript definitions for LightStep tracer specific options/methods
* Fixes a crash when `proto` transport was used and invalid responses were received.
* Adds an option to exclude cookies from tracing payloads.

## 0.22.0

* The default transport has been changed to `proto`.
* The tracer supports empty access tokens.
* Several bugs have been fixed with the `proto` transport option around the report of LightStep-specific tags.
* Addresses an issue with `proto` transport usage in the browser.
* Fixes bug where error objects are not converted to log objects correctly.
* Implements plugin for fetch.
* Remove deprecated chrome.loadTimes() code.

## 0.21.1

* Passing fractional timestamps to span start/finish are now properly handled.
* Several bugs were fixed with the `proto` transport option around span duration and span ids.
* We now support a custom logger function (thanks @Rowno), see README for more details.

## 0.21.0

* Update `thrift` to `0.11.0`. This breaks support for node.js versions <= 5.8.

## 0.20.14

* Add new option to Tracer: `disable_clock_skew_correction` `bool` *optional, defaults to 'false'*, which disables client-side clock kew correction.

## 0.20.13

* Add new option to Tracer: `transport` `string` *optional, defaults to 'thrift'*, which controls the transport method used to send reports to LightStep. *Experimental Feature*
* Fixes a bug when logging verbosity was greater than 4 during report flushes.

## 0.20.9

* Add new option to Tracer: `collector_path` `string` *optional*, which is an optional custom collector base path (if served behind a reverse proxy)

## 0.20.6

* `options` function now only logs warnings when it sees a key name it doesn't know. Previously, it threw an error.

## 0.20.6

* `instrument_page_load` option now defaults to false.
* If `xhr_instrumentation` is enabled, Open Tracing headers are now properly sent with every xhr request, so long as that request is to a domain that matches the whitelist url rules and also does not match any of the blacklist url rules.

## 0.20.0

* **API CHANGE**: use `new lightstep.Tracer(opts)` to return a lightstep `Tracer` instance
* **API CHANGE**: `lightstep-browser.min.js` and `lightstep-browser.js` use a `lightstep` rather than `LightStep` prefix for LightStep symbols
* **API CHANGE**: two new length limits for key-value logs: `log_field_key_hard_limit` and `log_field_value_hard_limit`, both specified as a number of bytes
* Many implementation changes, especially support for key-value logging per OpenTracing-javascript `0.13.*`

## 0.11.17

* Update the internal, automatic tracer tags to follow the convention of using a `lightstep.*` prefix


index.d.ts

/** Original definitions by felixfbecker: https://github.com/lightstep/lightstep-tracer-javascript/issues/99#issuecomment-436823685 */
declare module 'lightstep-tracer' {
  import * as opentracing from 'opentracing';

  export interface TracerOptions {
    /** the project access token. Access tokens are used by the LightStep tracer client libraries to identify the project the tracer is reporting for */
    access_token: string

    /** the string identifier for the application, service, or process. A Component is a logical service (or client) in a distributed system. The component usually represents a particular process or script in the distributed system */
    component_name: string

    /**
     * controls the level of logging to the console
     *
     * - 0 - the client library will never log to the console
     * - 1 - error reporting will be throttled to the first error per minute
     * - 2 - all errors are logged to the console
     * - 3 - all errors, warnings, and info statements are logged to the console
     * - 4 - all log statements, including debugging details
     *
     * @default 1
     */
    verbosity?: number

    /** custom collector hostname */
    collector_host?: string

    /** custom collector port */
    collector_port?: number

    /** custom collector base path (if served behind a reverse proxy) */
    collector_path?: string

    /** optional, default='tls'
     * `tls` - use HTTPS encrypted connections
     * `none` - use HTTP plain-text connections
    */
    collector_encryption?: "tls" | "none"

    /**
     * optional tag object that will be applied to all reports.
     */
    tags?: any

    /**
     * optional, the maximum time before the tracer reports. default = 2500
     */
    max_reporting_interval_millis?: number

    /**
     * optional, disables clock skew correction. default = false
     */
    disable_clock_skew_correction?: boolean

    /**
     * optional, specifies the transport method used for communication to the Satellite. default = thrift.
     * `thrift` - use thrift
     */
    transport?: string

    /**
     * optional, disables the tracer. default = false
     */
    disabled?: boolean

    /**
     * optional, defines the maximum number of spans per report. default = 4096
     */
    max_span_records?: number

    /**
     * optional tag object that will be applied to all spans.
     */
    default_span_tags?: any

    /**
     * optional, sets the timeout for flushing reports to Satellite. default = 30000 (30s)
     */
    report_timeout_millis?: number

    /**
     * optional, if enabled will use gzip compression of reports from node.js. default = true
     */
    gzip_json_requests?: boolean

    /**
     * optional, if enabled the tracer will not flush reports to Satellite. default = false
     */
    disable_reporting_loop?: boolean

    /**
     * optional, if enabled the tracer will not attempt to flush at process exit. default = false
     */
    disable_report_on_exit?: boolean

    /**
     * optional, specifies how long the tracer should wait before its first flush in ms. default = 1000
     */
    delay_initial_report_millis?: number

    /**
     * optional, specifies how quickly error logging should throttle in ms. default = 60000
     */
    error_throttle_millis?: number

    /**
     * optional, an override of the existing logging function for the tracer. by default, log to console.
     */
    logger?: any

    /**
     * optional. browser-only. creates a single long-lived span for the entire page view. default = false
     */
    instrument_page_load?: boolean

    /**
     * optional. browser-only. if enabled, automatically instruments all XHR requests with context headers.
     * see `xhr_url_inclusion_patterns` and `xhr_url_exclusion_patterns`.
     */
    xhr_instrumentation?: boolean

    /**
     * optional. browser-only. a regex to indicate which URLs should be automatically instrumented for XHRs.
     * default value is all urls.
     */
    xhr_url_inclusion_patterns?: RegExp[]

    /**
     * optional. browser-only. a regex to indicate which URLs should not be automatically instrumented for XHRs.
     * default value is no urls.
     */
    xhr_url_exclusion_patterns?: RegExp[]

    /**
     * optional. browser-only. a regex to indicate which automatically instrumented XHR URLs should include headers for continuing the trace on the server.
     * default value is all urls.
     */
    xhr_url_header_inclusion_patterns?: RegExp[]

    /**
     * optional. browser-only. a regex to indicate which automatically instrumented XHR URLs should not include headers for continuing the trace on the server.
     * default value is no urls.
     */
    xhr_url_header_exclusion_patterns?: RegExp[]

    /**
     * optional. browser-only. if enabled, automatically instrument all window.fetch requests with context.headers.
     * see `fetch_url_inclusion_patterns` and `fetch_url_exclusion_patterns`.
     */
    fetch_instrumentation?: boolean

    /**
     * optional. browser-only. a regex to indicate which URLs should be automatically instrumented for window.fetch.
     * default value is all urls.
     */
    fetch_url_inclusion_patterns?: RegExp[]

    /**
     * optional. browser-only. a regex to indicate which URLs should not be automatically instrumented for window.fetch.
     * default value is no urls.
     */
    fetch_url_exclusion_patterns?: RegExp[]

    /**
     * optional. browser-only. a regex to indicate which automatically instrumented URLs should include headers for continuing the trace on the server.
     * default value is all urls.
     */
    fetch_url_header_inclusion_patterns?: RegExp[]

    /**
     * optional. browser-only. a regex to indicate which automatically instrumented URLs should not include headers for continuing the trace on the server.
     * default value is no urls.
     */
    fetch_url_header_exclusion_patterns?: RegExp[]

    /**
     * optional. node-only. if enabled, automatically instrument all node requests with
     * context.headers.
     */
    nodejs_instrumentation?: boolean

    /**
     * optional. node-only. a regex to indicate which URLs should be automatically instrumented for
     * node's native http, https modules.
     * default value is all urls.
     */
    nodejs_url_inclusion_patterns?: RegExp[]

    /**
     * optional. node-only. a regex to indicate which URLs should not be automatically instrumented
     * for node's native http, https modules.
     * default value is no urls.
     */
    nodejs_url_exclusion_patterns?: RegExp[]

    /**
     * optional. browser-only. if true, includes cookies in the span logs for both `window.fetch` and `XMLHttpRequest`.
     * defaults to true.
     */
    include_cookies?: boolean
  }

  export class Tracer extends opentracing.Tracer {
    constructor(options: TracerOptions)
  }

}

